### ISSUE TITLE
[ttnn/prod]: float32 prod multiplies at ~10-bit precision: 3–4% error over 32 elements, 99% over 4096, product of 72,192 values near 1.0 returns 3.9e-31 (answer 98.9)

### ISSUE THREAD

**Issue by rybruscoe** [tenstorrent/tt-metal#57634]
### Component / Area

ops (ttnn/reduction/prod)

### Issue Type (optional)

Bad Outputs

### Observed

`ttnn.prod` documents `FLOAT32` input/output, and its program factories enable a 32-bit DEST for it. The product is nevertheless formed at roughly 10 bits per multiply with a downward bias, so it compounds to nothing. Inputs `exp(0.01·N(0,1))` (every value within 4% of 1.0, product well inside fp32 range), `[1, 1, 32, n]` float32, TILE, `dim=-1`; reference = exp(Σ log x) in float64; `torch.prod` float32 shown for scale.

| n | device rel error max | median | torch fp32 rel max | row 0: device | exact |
|---|---|---|---|---|---|
| 32 | **4.2e-2** | 3.1e-2 | 2.9e-7 | 0.9365 | 0.9670 |
| 1,024 | **6.7e-1** | 6.5e-1 | 3.2e-6 | 0.2945 | 0.8268 |
| 4,096 | **9.9e-1** | 9.9e-1 | 3.6e-6 | 0.0280 | 1.9621 |
| 72,192 | 1.0 | 1.0 | 1.6e-5 | **3.89e-31** | **98.89** |

Same on `dim=-2` (rel max 0.69 at 1024) and, smaller, on the NC path (`dim=1`, four factors: 7e-3). `prod()` over all 131,072 elements returns `1.4e-45` (a denormal) for a true value of order 1. `bfloat16` input is worse still (0.83 median rel error at n=1024), but it makes no precision claim.

The loss is rounding, not a wrong algorithm: exact-in-bf16 inputs are exact (all ones → 1; alternating 2, 0.5 → 1), and a factor that needs 10 mantissa bits is dropped entirely:

| input row (64 values) | device | exact |
|---|---|---|
| 32 × (1+2⁻¹⁰), then 32 × 1.0 | **1.0000000000** | 1.0317 |
| 32 × 1.0, then 32 × (1+2⁻¹⁰) | 1.0009765625 (one factor survives) | 1.0317 |
| 64 × (1+2⁻¹⁰) | 1.0009765625 | 1.0645 |
| 32 × (1+2⁻⁸), then 32 × 1.0 | 1.1250000000 | 1.1329 |

So each factor of `1+2⁻¹⁰` multiplies the running product by 1.0; the per-step relative error on the random inputs is −1.0e-3 (log(device/exact)/n at n = 32 and 1024), i.e. truncation to ~10 bits.

## Cause (two places, both in the prod data path; `is_fp32_dest_acc_en` is on)

1. **The input tile is unpacked to SrcA, not to DEST.** Both factories set, for a Float32 input, `compute_hw.unpack_modes.insert({INPUT_DFB, UnpackMode::UnpackToSrc})` with the comment *"Legacy set no unpack_to_dest_mode (default -> UnpackToSrc)"* (`prod_nc_program_factory.cpp:129-131`, `prod_all_program_factory.cpp:80-82`, `main` e4a9afc5). SrcA holds TF32 (10-bit mantissa), so every factor arrives at 10 bits. This is the mechanism #45222 described for std/var ("FPU SrcA reads cb_in at TF32 ... regardless of any precision-preservation plumbing downstream") and #45319 fixed there.
2. **The running product is re-read from DEST into SrcA every step.** `prod_nc.cpp` / `prod_all.cpp` keep the product in DEST and fold each tile with `mul_reuse_dest_tiles<DEST_TO_SRCA>`, which moves the fp32 partial product through SrcA again (the DEST→SrcA move is a 16-bit-high path; see the "MOVB2D/D2B depends on SrcA ALU Format - Hi/Lo16 does not work with Tf32" notes in `reduce.h`). That is why a factor of 1+2⁻¹⁰ vanishes even when the input side is exact.

### Expected

ttnn.prod on float32 input forms the product at float32 precision, so the result matches torch.prod in float32 within float32 sequential rounding (relative error ~1e-6 at n=1024, ~2e-5 at n=72,192), and a factor of 1+2^-10 is never dropped.

### 1. Steps (exact commands)

Run the snippet below on a Blackhole device with the listed tt-metal commit:

```python
import torch, ttnn
dev = ttnn.open_device(device_id=0)
torch.manual_seed(0)
x = torch.exp(0.01 * torch.randn(1, 1, 32, 4096))            # every factor within 4 % of 1
a = ttnn.from_torch(x, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=dev)
print(ttnn.to_torch(ttnn.prod(a, dim=-1, keepdim=True))[0, 0, 0, 0].item(), x.double().log().sum(-1).exp()[0, 0, 0].item())
# 0.0279986  vs  1.96207
```

### 2. Input data / link or description

Generated in the snippet above (torch.randn with manual_seed(0), exponentiated); no external data. The tables above were produced by the same script at the listed shapes and the listed constant rows.

### 3. Frequency

Always; deterministic across runs.

### 1. Software Versions

OS version: Ubuntu 22.04
Software / firmware release version: tt-metal a1c3bb93e (2026-09-16) with an unrelated cumsum patch; the prod factories and kernels are identical on main e4a9afc5 (2026-09-23).

### 2. Hardware Details

Product: Blackhole
Card/System: P150, single device

### Is this a regression?

No

### Logs & Diagnostics

Fix for both places, validated on Blackhole on the same inputs, with regression tests:

| n | rel error max (fixed) | median | torch fp32 rel max |
|---|---|---|---|
| 32 | 4.0e-7 | 7.6e-8 | 2.9e-7 |
| 1,024 | 1.6e-6 | 4.2e-7 | 3.2e-6 |
| 4,096 | 4.4e-6 | 9.0e-7 | 3.6e-6 |
| 72,192 | 2.0e-5 | 5.3e-6 | 1.6e-5 |

`dim=-2`: 3.4e-6 max; NC path: 1.8e-7; `(1+2⁻¹⁰)^64` = 1.0644619465 (exact to 5e-8); `prod()` over all elements returns an O(1) value instead of a denormal. Wall time unchanged (2.62 → 2.62 ms at [1,1,1024,4096]; 14.67 → 14.68 ms at [1,8,4096,4096]). The result then carries float32's own sequential rounding, the same as `torch.prod` in float32. Happy to open the PR.

### Priority

P2

### Impact

Affected workflows: any float32 prod over more than a few dozen elements is unusable: 67% error at 1,024 elements, 99% at 4,096, and a denormal for a product of order 1 over the whole tensor.
Release or date risk: none known.
